delivery_auto_refresh/models/sale_order.py

```diff
@@ -0,0 +1,110 @@
+                if rec.partner_id.property_delivery_carrier_id:
+                    rec.carrier_id = rec.partner_id.property_delivery_carrier_id
+                else:
+                    carrier_ids = self.env["delivery.carrier"].search(
+                        [
+                            "|",
+                            ("company_id", "=", False),
+                            ("company_id", "=", self.company_id.id),
+                        ]
+                    )
+                    rec.carrier_id = min(carrier_ids, key=lambda c: c.sequence)
+
+    def _get_param_value(self):
+        get_param = self.env["ir.config_parameter"].sudo().get_param
+        param = "delivery_auto_refresh.auto_add_delivery_line"
+        return safe_eval(get_param(param, "0"))
+
+    def _auto_refresh_delivery(self):
+        self.ensure_one()
+        # Make sure that if you have removed the carrier, the line is gone
+        if self.state in {"draft", "sent"}:
+            self._remove_delivery_line()
+        if self._get_param_value() and self.carrier_id:
+            if self.state in {"draft", "sent"} or self.invoice_shipping_on_delivery:
+                price_unit = self.carrier_id.rate_shipment(self)["price"]
+                self._create_delivery_line(self.carrier_id, price_unit)
+
+    @api.model
+    def create(self, vals):
+        """Create or refresh delivery line on create."""
+        order = super().create(vals)
+        order._auto_refresh_delivery()
+        return order
+
+    def write(self, vals):
+        """Create or refresh delivery line after saving."""
+        res = super().write(vals)
+        if self._get_param_value():
+            for order in self:
+                delivery_line = order.order_line.filtered("is_delivery")
+                if len(delivery_line) > 1:
+                    continue
+                order.with_context(
+                    delivery_discount=delivery_line.discount,
+                )._auto_refresh_delivery()
+        return res
+
+    def _create_delivery_line(self, carrier, price_unit):
+        """Allow users to keep discounts to delivery lines. Unit price will
+           be recomputed anyway"""
+        sol = super()._create_delivery_line(carrier, price_unit)
+        discount = self.env.context.get("delivery_discount")
+        if discount and sol:
+            sol.discount = discount
+        return sol
+
+    def _action_confirm(self):
+        res = super()._action_confirm()
+        for so in self:
+            so.invoice_shipping_on_delivery = all(
+                [not line.is_delivery for line in so.order_line]
+            )
+        return res
+
+    def action_open_delivery_wizard(self):
+        if self._get_param_value():
+            raise UserError(
+                _(
+                    "The shipping cost is updated automatically by changing "
+                    "the Delivery Method field."
+                )
+            )
+        super().action_open_delivery_wizard()
```

Comment:
I don't know if this is normal but, while on draft and with the parameter set to False, if I press this button nothing happens in the UI. Is it the intended behavior?